standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/RetryingHMSHandler.java

```diff
@@ -227,6 +231,26 @@ public Result invokeInternal(final Object proxy, final Method method, final Obje
     }
   }
 
+  private Class<SQLException>[] unrecoverableSqlExceptions = new Class[]{
+          // TODO: collect more unrecoverable SQLExceptions
+          SQLIntegrityConstraintViolationException.class
+  };
+
+  private boolean isRecoverableException(Throwable t) {
+    assert t != null && (t instanceof JDOException || t instanceof NucleusException);
```

Comment:
note, that assertions won't get executed unless you enable them explicitly `-enableassertions` (-ea)